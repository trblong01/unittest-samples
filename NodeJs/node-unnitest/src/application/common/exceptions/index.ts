import { ZodError } from 'zod';

export class ValidationException extends Error {
  constructor(error: ZodError) {
    super('APPLICATION_VALIDATION_ERROR');
    this.name = 'ValidationException';
    this.errors = error.issues.map((issue) => ({
      path: issue.path.join('.'),
      message: issue.message,
    }));
  }

  errors: { path: string; message: string }[];
}
