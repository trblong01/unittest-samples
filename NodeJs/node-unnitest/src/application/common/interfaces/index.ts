export * from './config';
export * from './logger';
export * from './posts-repository';
