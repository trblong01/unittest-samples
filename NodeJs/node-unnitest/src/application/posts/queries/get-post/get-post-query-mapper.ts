import { Post } from '@domain/entities';

export function toDto(post: Post) {
  return {
    id: post.id,
    title: post.title,
  };
}
