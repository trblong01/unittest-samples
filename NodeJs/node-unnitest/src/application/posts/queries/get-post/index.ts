export * from './get-post-query';
