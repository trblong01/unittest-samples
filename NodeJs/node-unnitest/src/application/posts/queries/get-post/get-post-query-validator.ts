import { ValidationException } from '@application/common/exceptions';
import { ZodError, z } from 'zod';

import { GetPostQuery } from './get-post-query';

export async function validate(query: GetPostQuery) {
  try {
    const schema: z.ZodType<GetPostQuery> = z.object({
      id: z.string().uuid(),
    });

    await schema.parseAsync(query);
  } catch (error) {
    throw new ValidationException(error as ZodError);
  }
}
