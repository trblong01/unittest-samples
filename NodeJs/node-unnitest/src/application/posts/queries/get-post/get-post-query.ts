import { toDto } from './get-post-query-mapper';
import { validate } from './get-post-query-validator';
import axios from 'axios';

export type GetPostQuery = Readonly<{
  id: string;
}>;

export function makeGetPostQuery({ postsRepository }: Pick<Dependencies, 'postsRepository'>) {
  return async function getPostQuery({ id }: GetPostQuery) {
    await validate({ id });

     // Make an HTTP call to the external API
     const externalApiUrl = `https://external-system.example.com/author/${id}`;
     const externalApiResponse = await axios.get(externalApiUrl);
 
     const { authorName, authorDescription } = externalApiResponse.data;
 
     // Validate the external API response
     if (!externalApiResponse.data || !authorName || !authorDescription) {
       throw new Error('Invalid author data received from the external system.');
     }

    const post = await postsRepository.getById({ id });

    if (!post) {
      return null;
    }

    return toDto(post);
  };
}
