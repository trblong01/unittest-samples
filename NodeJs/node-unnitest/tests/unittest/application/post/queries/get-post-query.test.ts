
// import { PostsRepository } from '@application/common/interfaces/posts-repository';
// import { toDto } from '@application/posts/queries/get-post/get-post-query-mapper';
import { ValidationException } from '@application/common/exceptions';
import { validate } from '@application/posts/queries/get-post/get-post-query-validator';
// import { Post } from '@domain/entities/post';
import axios from 'axios';

import { makeGetPostQuery } from '../../../../../src/application/posts/queries/get-post/get-post-query';
jest.mock('axios');
// jest.mock('../../../../../src/application/posts/queries/get-post/get-post-query-validator',() => ({
//   validate: jest.fn(),
// }));
// jest.mock('../../../../../src/application/posts/queries/get-post-query-mapper',() => ({
//   toDto: jest.fn(),
// }));

describe('makeGetPostQuery', () => {
  const postsRepository = {
    getById: jest.fn(),
    create: jest.fn(), // Mocking the 'create' method
    delete: jest.fn(), // Mocking the 'delete' method
    list: jest.fn(),   // Mocking the 'list' method
  };

  
  const getPostQuery = makeGetPostQuery({ postsRepository });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Validation', () => {
    it('should validate the input', async () => {
      const id = '123';
      (axios.get as jest.Mock).mockResolvedValue({ data: { authorName: 'John Doe', authorDescription: 'Author Description' } });
      postsRepository.getById.mockResolvedValue({});

      expect(getPostQuery({ id })).rejects.toThrow(ValidationException);
    });
  });

  describe('External API Call', () => {
    it('should throw an error if external API returns invalid data', async () => {
      const id = '907b0a0b-9a12-4073-ab27-3ad5927955e9';
      (axios.get as jest.Mock).mockResolvedValue({ data: { authorName: null } });

      await expect(getPostQuery({ id })).rejects.toThrow('Invalid author data received from the external system.');
    });
  });

  describe('Repository Interaction', () => {
    it('should return null if the post does not exist', async () => {
      const id = '907b0a0b-9a12-4073-ab27-3ad5927955e9';
      (axios.get as jest.Mock).mockResolvedValue({ data: { authorName: 'John Doe', authorDescription: 'Author Description' } });
      postsRepository.getById.mockResolvedValue(null);

      const result = await getPostQuery({ id });

      expect(result).toBeNull();
    });

    it('should return the post DTO if the post exists', async () => {
      const id = '907b0a0b-9a12-4073-ab27-3ad5927955e9';
      const post = { id: '907b0a0b-9a12-4073-ab27-3ad5927955e9', title: 'Post Title' };
      // const postDto = new Post({ id: '907b0a0b-9a12-4073-ab27-3ad5927955e9', createdAt: new Date(2022, 1, 1), title: 'Post Title DTO'  });

      (axios.get as jest.Mock).mockResolvedValue({ data: { authorName: 'John Doe', authorDescription: 'Author Description' } });
      postsRepository.getById.mockResolvedValue(post);
      // (toDto.mockReturnValue(postDto);

      const result = await getPostQuery({ id });

      expect(result).toEqual(post);
      
    });
  });
});