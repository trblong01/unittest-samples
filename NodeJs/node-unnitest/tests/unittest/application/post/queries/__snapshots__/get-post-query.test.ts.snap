// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPostQuery given the post exists should return the post 1`] = `
{
  "id": "907b0a0b-9a12-4073-ab27-3ad5927955e9",
  "title": "Mock post",
}
`;
